### PR TITLE
Full Project Implementation!

### PR DESCRIPTION
**Pull request by xarrijorge**
Complete testing of chosen enum methods.

### CODE REVIEW

File: spec/custom_enum_spec.rb [hunk 1]
```diff
@@ -0,0 +1,112 @@
+require './enum_methods/custom_enum.rb'
+
+RSpec.describe Enumerable do
+    describe "#my_each" do
+      it " It returns each element in array when given a block " do
+        array=[1,2,3,4,5]
```
Comment (mcrd25) [inline spec/custom_enum_spec.rb]: Try using let statement to declare test variables

For example after line 3 you'd have:
```ruby
let(:test_array) { [1, 2, 3, 4, 5] }
```

This way you can also use that same test array in multiple tests

File: spec/custom_enum_spec.rb [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+        expect( array.my_each { |a|  a } ).to eql([1,2,3,4,5])
```
Comment (mcrd25) [inline spec/custom_enum_spec.rb]: Because you'd have declared the array could have statement be like

`expect( test_array.my_each { |a|  a } ).to eql(test_array)`

Try to avoid hard-coded values. That way you'd only have to change one variable instead of all occurrences of it... Especially in tests where you'd test same array multiple times

File: spec/custom_enum_spec.rb [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+      end
+      
+      it " It returns each key or value in hash when given a block with two parameter " do
+        hash={ 1 => 'hi', 2 => 'hello', 3 => 'how'}
```
Comment (mcrd25) [inline spec/custom_enum_spec.rb]: See my previous comment and apply accordingly

File: spec/custom_enum_spec.rb [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+        expect( hash.my_each { |a,b|  b } ).to eql( {1=>"hi", 2=>"hello", 3=>"how"} )
+      end
+
+    end
+
+    describe " #my_count(para = nil) " do 
+      it "It returns the number of elements in array when no parameter is given and no block is passed " do
+        array = [1,2,3,4,5]
```
Comment (mcrd25) [inline spec/custom_enum_spec.rb]: With let would not have to do this

File: spec/custom_enum_spec.rb [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+         expect( array.my_count ).to eql( 5 )
+      end
+
+      it "It returns the number of the argument that exist in the array it operates upon,when no block is given " do
+        array=['a','a','a','b']
```
Comment (mcrd25) [inline spec/custom_enum_spec.rb]: You cant name this test_array2 or array2 after you have declared you first array after line 3... so after line 3 you'd have something like:
```ruby
let(:test_array) { [1, 2, 3, 4, 5] }
let(:test_array2) { ['a', 'a', 'a', 'b'] }
```

File: spec/custom_enum_spec.rb [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+         expect( array.my_count('a') ).to eql( 3 )
+      end
+       
+      it "It returns the number of that exist in the array based on the condition specified in the block " do
+        array=['a','a','a','b']
```
Comment (mcrd25) [inline spec/custom_enum_spec.rb]: Would not need to have to declare this again as by now you'd have it declared in a suitable let variable before the tests

File: spec/custom_enum_spec.rb [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+         expect( array.my_count { |a| a  =='b' } ).to eql( 1 )
+      end
+      
+    end
+
+    describe "#my_any?(arg = nil)" do
+       it "It return false if array is empty and neither a block or agument is passed" do
+        array = []
```
Comment (mcrd25) [inline spec/custom_enum_spec.rb]: you can declare an empty array in the manner mentioned previously and name variable empty_array etc.

File: spec/custom_enum_spec.rb [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+        expect( array.my_any? ).to eql(false)
+       end
+       it "It return true if any element in the array is true and neither a block or argument is given" do
+        array = [false, false, true]
```
Comment (mcrd25) [inline spec/custom_enum_spec.rb]: At this point I think you get the gist of what I'm saying so apply to all occurrences

File: spec/custom_enum_spec.rb [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+
+       it "It return true based on the codition specified in the block when give a block" do
+        array = ['kofi','ama', 'adwoa']                     
+        expect( array.my_any? { |word| word.length >= 3 }  ).to eql(true)
+       end
+
+       
+       it "It return false if none of the above condition is not true" do
+        array = ['kofi','akosua', 'adwoa']                     
+        expect( array.my_any? { |word| word.length == 3 }  ).to eql(false)
```
Comment (mcrd25) [inline spec/custom_enum_spec.rb]: You can actually create a block like this
`let(:test_block) { proc { |word| word.length == 3 } }`

**Review by mcrd25 — CHANGES_REQUESTED** [review]
# Great attempt! You did a great job with methods and their respective tests! 😄 

**Just a couple of things to adjust and some suggestions**
- Address review comments (apply to all cases where fits) (required)
- you did not do tests for the following methods: (required)
  - my_all?
  - my_none
  - my_each_with_index

- Try to avoid hard-coded values so say you have an array you'd like to use to test you can have something like 
  ```ruby
  ARRAY_SIZE = 100
  LOWEST_VALUE = 0
  HIGHEST_VALUE = 9
   let(:array) { Array.new(ARRAY_SIZE) { rand(LOWEST_VALUE...HIGHEST_VALUE) } }
  ```
  (suggested)
  That way you'd a randomly generated array that doesn't necessarily have same values and you get a better idea that your method works for various values. Not just values you think of directly.
- try testing for if it's eq to the original. So something like this 
  `expect(test_arr.my_each { test_block }).to eq(test_arr.each { test_block })` (suggested)

**We are almost there 👍. Make changes and request another code review**
 
***If you have any questions, let me know here or on slack and I will be glad to clarify with you. You can also reach me and connect with me on my social media.***

**[--Maya](https://github.com/mcrd25)**
[GitHub](https://github.com/mcrd25) [Twitter](https://twitter.com/mcrd25) [linkedin](https://www.linkedin.com/in/mayadouglas/)

File: spec/custom_enum_spec.rb (hunk 3 above)
Comment (Forison) [reply spec/custom_enum_spec.rb]: hi 
per the requirement of the project we are suppose to test any six of the methods not all.
Again the method my_any? was duely tested ...

File: spec/custom_enum_spec.rb (hunk 1 above)
Comment (xarrijorge) [reply spec/custom_enum_spec.rb]: `Go back to the Advanced Building Blocks Project and write tests for any 6 of the enumerable methods you wrote there. In this case, identify several possible inputs for each of those functions and test to make sure that your implementation of them actually makes the tests pass. Be sure to try and cover some of the odd edge cases where you can.`

The project says any six so that we chose those six to submit. Could you please point us to where the above methods are mentioned as required?

File: spec/custom_enum_spec.rb (hunk 7 above)
Comment (xarrijorge) [reply spec/custom_enum_spec.rb]: But we would be pushing things to that array later, wouldn't the name "empty_array" be wrong during the return?

Comment (mcrd25) [reply spec/custom_enum_spec.rb]: As I can see from the test you are not. This particular array is an empty array...
Remember these are tests.

I see that you are assigning new values to the array but you are not pushing anything to it unless I'm mistaken?

File: spec/custom_enum_spec.rb (hunk 1 above)
Comment (mcrd25) [reply spec/custom_enum_spec.rb]: Ahhh yes, the odin project states this but Microverse requires you to test all methods implemented... Ideally though if you are doing TDD you really should test all public methods no?

Honestly, I understand your confusion and hope that you will implement all tests

It would benefit you to do so.

Here are my guidelines:

### Project Specification Shortcut

- [ ] All the methods implemented in the Enumerables project are tested
- [x] Check if specs are passing on your machine.
- [x] Check if specs are failing if you break some code in tested class.

### Crucial things to make project works

- [x] The project has a descriptive README.md
- [x] Verify if the project should be done solo or collaboratively and that the commit history reflects that
- [x] Comments: It’s better to delete commented code. If you need to review a previous version of your code, use git history.


- [x] _IF NOT CAUGHT BY RUBOCOP_ make sure that code is clean: indentation with two spaces, line lengths, spaces around operators - first section of https://github.com/github/rubocop-github/blob/master/STYLEGUIDE.md#ruby-style-guide
- [ ] A bit of common sense: check if specs cover positive and negative scenarios 
      Example: there is a spec for correct usage of method [e.g. “it “creates object if all required attributes are passed”]
      and also there is spec for wrong usage of method  [e.g. “it “does not create object if required attributes are missing”].
- [ ] Specs for each method should be wrapped in one `describe` block. As in example here https://github.com/rubocop-hq/rspec-style-guide#empty-lines-around-examples .
- [ ] You should have one `it` block for each expected behaviour.
- [ ] Instead of using `before_each` to define a variable for the test cases, students should [Use `let` instead](https://github.com/rubocop-hq/rspec-style-guide#let-blocks).


- [ ] All the methods implemented in the Enumerables project are tested

**Review by mcrd25 — CHANGES_REQUESTED** [review]
# Great job again! 3 more methods to test and you'd be able to 🚀 

Please do tests for following methods:
- my_all?
- my_each_with_index
- my_none?

**Everything looks great so far! Keep it up! 💯**

***If you have any questions, let me know here or on slack and I will be glad to clarify with you. You can also reach me and connect with me on my social media.***

**[--Maya](https://github.com/mcrd25)**
[GitHub](https://github.com/mcrd25) [Twitter](https://twitter.com/mcrd25) [linkedin](https://www.linkedin.com/in/mayadouglas/)

**Review by JamesOkunlade — APPROVED** [review]
### Project well did!!

I see it's a long way down here, you've have written completely all the tests as required of you by the previous TSE. This is good to pass now.

You should merge dev to master, close pull request and fill the project completion form.

Good luck with your future project. :+1::+1:
